src/factom-event-emitter.js

```diff
@@ -0,0 +1,274 @@
+        try {
+            const factoidBlock = await this._cli.getFactoidBlock(directoryBlock.factoidBlockRef);
+
+            if (this.listenerCount(FACTOM_EVENT.factoidBlock)) {
+                this.emit(FACTOM_EVENT.factoidBlock, factoidBlock);
+            }
+
+            if (this._factoidAddressSubscriptions.size) {
+                // Must pass directoryBlock through to access block context
+                this._emitFactoidTransaction(factoidBlock, directoryBlock);
+            }
+        } catch (err) {
+            this.emit('error', err);
+        }
+    }
+
+    async _emitAdminBlock(directoryBlock) {
+        try {
+            const adminBlock = await this._cli.getAdminBlock(directoryBlock.adminBlockRef);
+            this.emit(FACTOM_EVENT.adminBlock, adminBlock);
+        } catch (err) {
+            this.emit('error', err);
+        }
+    }
+
+    async _emitEntryCreditBlock(directoryBlock) {
+        try {
+            const entryCreditBlock = await this._cli.getEntryCreditBlock(directoryBlock.entryCreditBlockRef);
+            this.emit(FACTOM_EVENT.entryCreditBlock, entryCreditBlock);
+        } catch (err) {
+            this.emit('error', err);
+        }
+    }
+
+    // Emit the first entry block of any newly created entry chain.
+    async _emitNewEntryChains(directoryBlock) {
+        try {
+            const checkIfNewChainAndEmit = async ref => {
+                const entryBlock = await this._cli.getEntryBlock(ref.keyMR);
+                if (entryBlock.sequenceNumber === 0) {
+                    this.emit(FACTOM_EVENT.newEntryChain, entryBlock);
+                }
+            };
+
+            // Fetch entry chains contained in directory block concurrently
+            await Promise.map(directoryBlock.entryBlockRefs, checkIfNewChainAndEmit, { concurrency: 5 });
+        } catch (err) {
+            this.emit('error', err);
+        }
+    }
+
+    // Emit the latest entry block of any entry chain the user is listening to.
+    async _emitEntryBlock(directoryBlock) {
+        const fetchAndEmitNewBlock = async chainId => {
+            try {
+                const entryBlockRef = directoryBlock.entryBlockRefs.find(ref => ref.chainId === chainId);
+
+                if (entryBlockRef) {
+                    const entryBlock = await this._cli.getEntryBlock(entryBlockRef.keyMR);
+                    this.emit(chainId, entryBlock);
+                }
+            } catch (err) {
+                this.emit('error', err);
+            }
+        };
+
+        // Fetch new entry blocks concurrently
+        await Promise.map(this._entryChainSubscriptions, fetchAndEmitNewBlock, { concurrency: 5 });
+    }
+
+    // Emit new factoid transactions for user-defined addresses
+    _emitFactoidTransaction(factoidBlock, directoryBlock) {
+        const addrs = this._factoidAddressSubscriptions;
+        factoidBlock.transactions.forEach(tx => {
+            const activeAddresses = new Set();
```

Comment:
Just a suggested alternative:
`const activeAddresses = new Set([...tx.inputs, ...tx.factoidOutputs].filter(io => addrs.has(io.address)))`
I have a preference for it because it is more FP style (no side effects) and more concise

Reply:
It was necessary to also map over the result of filter in order to get the benefit of the Set.

```javascript
const activeAddresses = new Set([...tx.inputs, ...tx.factoidOutputs].filter(io => addrs.has(io.address).map(io => io.address)));
```